Ability to instrument a GraphQLArgument value

yarinvak opened the pull request:
I think that it's a nice functionality that can be added to the `GraphQLArgument`.
This PR adds the `GraphQLArgumentInstrumentation` interface.
A user can implement this interface, in order to make some transformation to the graphql argument's value, or to add a validation on top of it.

**So how it works?**
`GraphQLArgument` now contains a member called `instrumentation` that can be instantiated using the `GraphQLArgument.Builder`.
If you add an instrumentation object to the `GraphQLArgument` object, than the `ValuesResolver`, 
which calculate the value of the argument, will call the `GraphQLArgumentInstrumentation.instrumentValue` method, after calculating the value.
It will pass the value to the method, and the method can return a different value, or create some validation that will throw an exception if the value is bad.

**An example**
Lets say my field `name` has an argument `suffix`, so I can call it like that `name(suffix: "the best")` when executing my query.
Now I want to validate that the suffix only contains lower-case letters.

I can create the following `GraphQLArgumentInstrumentation`:
```java
public class LowerCaseValidation implements GraphQLArgumentInstrumentation{
   Object instrumentValue(Object argumentValue){
       if (argumentValue instanceof String && !isStringLowerCase(argumentValue)){
          throw new Exception("the argument value is not lower case!");
       }
       return argumentValue;
   }
}
```
Of course I can also use this class to transform the argument value to upper case (for example), so instead of "the best", its value will become "THE BEST".

How to add the instrumentation to the `GraphQLArgument`? easy: the builder has an `instrumentation` method that accepts `GraphQLArgumentInstrumentation` instance.

It is a kind of an idea I had in my head (I thought that directives may be used here but I'm not sure if that's their purpose).
I would be happy to change things if needed and to get other people thoughts about it.

bbakerman commented:
Thanks for this PR.   I will have a more detailed look at it soon.

This kind of capability has been much on my mind

bbakerman commented:
I have been thinking long and hard on this area.  The following is half musing and half feedback on this PR and on the general idea.

People in Java / JVM land want richer objects when they get their arguments to a field.  They want it for validation reasons and also for working with POJO reasons.

This prompted me to write https://github.com/graphql-java/graphql-java-extended-validation

This PR also makes moves in this area.  

The first bit of feedback I have is the naming.  I would not use Instruwentation - its already taken.

I suggest

* ArgumentConversion
* ArgumentConverter
* ArgumentMapper


Because I think a lot of people would put Jackson / GSON  whatever behind this and "map" from Map/List/primitive input into POJO land.

Let's use `ArgumentMapper` as a working name in terms of this feedback


 I would like to see more test code around more complex objects.  Today the tests and examples are on String arguments - which while common are also easy peasy.

What about complex input like this:

```
type Query{
     field ( arg : PersonInput) : String
}

input PersonInput {
    name : String
    age : Int
    parents : [PersonInput]
}

```

Can we get some tests for that to discover if it works well enough and is easy to use.

In the extended-validation code I spent a lot of time on SDL and use of directives.  I am bullish on the idea that people use SDL as a way to specify type systems and directives for possible behaviour

So how would this work with our new `ArgumentMapper` capability.  We would need something in RuntimeWiring code land to be able to provide ArgumentMappers.  Perhaps is best done by SchemaDirectiveWiring like I used in extended-validation

Musing allowed is there any sense that "argument mapping" can start lower in the chain that the top arguments to a field.  That is imagine

```
type Query{
     field ( arg : PersonInput) : String
}

input PersonInput {
    name : String
    age : Int
    parents : [PersonInput]
   hobbies : [Hobby]
}

@pojo(name : "package.name.Hobby")
input Hobby {
   name : String
   funFactor : Int
}
```

Is it even sensible to  suggest that we only want to convert from `Hobby` downwards?   Or would it be all or nothing for mapping conversions?   Or is this a choice of the mapper implementation?  These are just thoughts I have I wanted to put up

`ValueResolver` is used in more than one place right now.  I think we would need to  do a bit fo a code audit and make sure it makes sense in the other places it is called.  For example I think in argument validation ( the graphql kind not the business logic layer kind) we need the arguments to be in Map/List/primitive format so we can inspect them programmatically.  This needs to be checked.


Overall I am bullish on ths idea.

cc: @andimarek / @JaidenAshmore  - what are your thoughts here?

yarinvak commented:
Thanks for the feedback.
I'm ok with the mapper name, maybe even "transformer" can be used.
I'll add some more complex tests.

The problem in graphql-java & the extended validation, is that only those who use SDL in order to create their schema can use the directives. 
In my project graphql-java-annotations we do not use SDL but create the objects as simple POJO. 

I think that at first graphql-java should support the  wiring separately from the SDL. We had to create our functionality in graphql-java-annotations of runtime wiring in order to allow our users to use directives in order to wire their schemas.

JaidenAshmore commented on src/main/java/graphql/GraphQLArgumentInstrumentation.java:
```diff
@@ -0,0 +1,13 @@
+ * For example, if the client passes an argument value "hello" to an argument in your field, you can write an instrumentValue implementation that transforms it to upper case letters.
+ * Then, the data fetcher environment will contain a value of "HELLO".
+ */
+public interface GraphQLArgumentInstrumentation {
```
What are people's opinion of having this just be extending a `Function`. For example:

```
public interface GraphQLArgumentMapper extends Function<Object, Object> {

}
```
The one positive for this is it will allow for the consumer to apply chaining etc without having to build our own `ChainingGraphQLArgumentMapper`.  For example given the following mappers:

```
   public interface GraphQLArgumentMapper extends Function<Object, Object> {
        
    }
    
    public class UpperCaseGraphQLArgumentMapper implements GraphQLArgumentMapper {
        @Override
        public Object apply(final Object o) {
            if (o instanceof String) {
                return ((String) o).toUpperCase();
            }
            
            return o;
        }
    }

    public class StringPrefixGraphQLArgumentMapper implements GraphQLArgumentMapper {
        @Override
        public Object apply(final Object o) {
            if (o instanceof String) {
                return "prefix" + o;
            }

            return o;
        }
    }
```

then as a consumer you can then chain them like so:

``` 
        GraphQLArgumentMapper f = new UpperCaseGraphQLArgumentMapper();
        GraphQLArgumentMapper g = new StringPrefixGraphQLArgumentMapper();

        f.andThen(g)
                .apply("test");
```

yarinvak replied:
It's a nice idea. I think in this case the whole `GraphQLArgumentMapper` class is redundant and the type of the mapper can be just `Function`

JaidenAshmore commented on src/test/groovy/graphql/schema/GraphQLArgumentTest.groovy:
```diff
@@ -23,7 +36,7 @@ class GraphQLArgumentTest extends Specification {
                     .withDirective(newDirective().name("directive1"))
                     .withDirective(newDirective().name("directive3"))
-                    .value("VALUE")
+                    .value("VALUE").instrumentation(instrumentation2)
```
NIT: new line

JaidenAshmore commented on src/test/groovy/graphql/schema/GraphQLArgumentTest.groovy:
```diff
@@ -10,10 +11,22 @@ class GraphQLArgumentTest extends Specification {
 
     def "object can be transformed"() {
         given:
+        def instrumentation = new GraphQLArgumentInstrumentation() {
```
As brad said, some more complicated tests would be good too.

yarinvak commented:
You are welcome to review my new changes.
I added a test for a complex input type, I think the logic that happens here is pretty obvious and it should work for every complex type.
I fixed stuff according to your comments.
Hope it's better now :)

andimarek commented:
Thanks a lot @yarinvak for your PR. Unfortunately we don't think this PR takes the approach we want to go with.

- We decided to move away from Logic/Code/Callbacks inside GraphQL*** Objects. We invented the `GraphQLCodeRegistry` to separate these things further. 
- ValuesResolver should do exactly what it does today and not have an arbitrary Object->Object function in it.

These aspects could be fixed maybe, but even more importantly we are not sure yet how to deal with this requirement to map Arguments to Java Objects: DataFetcher is exactly right the way it is at the moment from an engine PO. Mixing it up with Java Object conversion (and mapping it to "nice" business methods) has a lot of consequences and we are not really sure graphql java is the right place to tackle these challenges. We are thinking this responsibility is better placed in libraries like Spring which already tackle these kind of things.

Thanks again for you RP: without contributors like you graphql java would not have been possible.

yarinvak commented:
Thanks @andimarek for your comment. What about including the mapper inside the code registry to maintain the separation of code and types?
Are you against the concept of the PR - allowing people to transform the graphql argument? Because I don't see a way of other libraries to transform the values of arguments in a generic manner before the argument arrives to the data fetcher (in a simple and nice way..).
It can also use for the purpose of argument directives - right now there is no way to wire the argument value with a directive, and I think it's a nice feature that can be usefull for many clients of graphql-java.
